### PR TITLE
Typing Op::TextDrawAdjusted (TJ)

### PR DESCRIPTION
**Pull request by markcatley**
Untyped arrays are a bit of a pain. Adding some extra type information to Op::TextDrawAdjusted (TJ)

### CODE REVIEW

**Comment by s3bk** [comment]
Does the spec require that spacing and strings alternate?
I could see some writers generating consecutive spaces or strings...

**Comment by s3bk** [comment]
> Show one or more text strings, allowing individual glyph positioning. Each element of array shall be either a string or a number. If the element is a string, this operator shall show the string. If it is a number, the operator shall adjust the text position by that amount; that is, it shall translate the text matrix, T .

**Comment by divergentdave** [comment]
I have a counterexample, there exists a PDF file where a TJ command ends with a spacing adjustment number. https://github.com/mozilla/pdf.js/issues/2808

**Comment by markcatley** [comment]
Sorry about that. I didn't realise. I've updated the implementation to allow this.

File: pdf/src/content.rs [hunk 1]
```diff
@@ -394,7 +394,18 @@ impl OpBuilder {
             }
             "Tf"  => push(Op::TextFont { name: name(&mut args)?, size: number(&mut args)? }),
             "Tj"  => push(Op::TextDraw { text: string(&mut args)? }),
-            "TJ"  => push(Op::TextDrawAdjusted { array: array(&mut args)? }),
+            "TJ"  => {
+                let mut primatives = array(&mut args)?.into_iter();
+                let mut array = Vec::<TextDrawAdjusted>::new();
+
+                while let Some(spacing_or_text) = primatives.next() {
```
Comment (s3bk) [inline pdf/src/content.rs:401]: looks like a or loop to me? `for spacing_or_text in primatives`

File: pdf/src/content.rs [hunk 2]
```diff
@@ -394,7 +394,18 @@ impl OpBuilder {
...
+                    let spacing_or_text = spacing_or_text.as_number().map(TextDrawAdjusted::Spacing).or_else(|_| spacing_or_text.into_string().map(TextDrawAdjusted::Text))?;
```
Comment (s3bk) [inline pdf/src/content.rs:402]: I think a match would be cleaner here:
```rust
match spacing_or_text {
    Primitive::Integer(i) => array.push(TextDrawAdjusted::Spacing(i as f32)),
    Primitive::Number(f) => array.push(TextDrawAdjusted::Spacing(f)),
    Primitive::String(s) => array.push(TextDrawAdjusted::Text(s)),
    p => bail!("invalid primitive in TJ operator: {:?}", p)
}
```

File: pdf/src/content.rs (hunk 1 above)
Comment (markcatley) [reply pdf/src/content.rs]: Whoops, old code leaking through.

File: pdf/src/content.rs (hunk 2 above)
Comment (markcatley) [reply pdf/src/content.rs]: Good idea.

**Review by s3bk — APPROVED** [review]

**Comment by s3bk** [comment]
Thanks!

**Pull request merged**